### PR TITLE
Added script for changelog.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Other

Short description (up to few sentences):
Added script which creates changelog from pull requests description.

Detailed description (optional):
[Raw result for last changelog](https://gist.github.com/KochetovNicolai/fe6ccc050d5a27f480d618878e2d0bb6).

### CODE REVIEW

File: utils/make_changelog.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+# Load inner state. For debug purposes.
+def load_state(state_file, base_sha, new_tag, prev_tag):
+
+    state = {}
+
+    if state_file:
+        try:
+            if os.path.exists(state_file):
+                logging.info('Reading state from {}'.format(state_file))
```
Comment (alesapin) [inline utils/make_changelog.py]: May use %s modifier in `logging.level("foo %s", "bar")` and get rid of `format`.

File: utils/make_changelog.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    text = 'Commits which are not from any pull request:\n\n'
+    return text + '\n\n'.join(texts)
+
+
+# List of pull requests -> text description.
+def process_pull_requests(pull_requests, users, repo):
+    groups = {}
+
+    for id, item in pull_requests.items():
+        lines = list(filter(len, map(lambda x: x.strip(), item['description'].split('\n'))))
```
Comment (alesapin) [inline utils/make_changelog.py]: list comprehension more preferable than `filter` and `map`, but it's ok.

File: utils/make_changelog.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+# Save inner state. For debug purposes.
+def save_state(state_file, state):
+    with codecs.open(state_file, 'w', encoding='utf-8') as f:
+        f.write(json.dumps(state, indent=4, separators=(',', ': ')))
+
+
+def make_changelog(new_tag, prev_tag, repo, repo_folder, state_file, token, max_retries, retry_timeout):
+
+    base_sha = get_merge_base(new_tag, prev_tag, repo_folder)
+    logging.info('Base sha: {}\n'.format(base_sha))
```
Comment (alesapin) [inline utils/make_changelog.py]: redundant new line?

File: utils/make_changelog.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+    prev_release_tag = args.prev_release_tag
+    new_release_tag = args.new_release_tag
+    token = args.token or ''
+    repo_folder = args.directory or SCRIPT_DIR
+    state_file = args.state
+    repo = args.repo
+    max_retry = args.max_retry
+    retry_timeout = args.retry_timeout
+
+    logging.basicConfig(level=logging.INFO)
```
Comment (alesapin) [inline utils/make_changelog.py]: `logging.basicConfig(level=logging.INFO, format='%(asctime)s %(message)s')` -- much better

File: utils/make_changelog.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+    logging.info('Found {} users.'.format(len(users)))
+    save_state(state_file, state)
+
+    print(process_pull_requests(pull_requests, users, repo))
+    print(process_unknown_commits(unknown_commits, commits_info, users))
+
+
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser(description='Make changelog.')
```
Comment (alesapin) [inline utils/make_changelog.py]: Need to add example usage in description or add `README.md`

File: utils/make_changelog.py [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+            resp = requests.get(url, headers={"Authorization": "token {}".format(token)})
+        else:
+            resp = requests.get(url)
+
+        if resp.status_code != 200:
+            msg = "Request {} failed with code {}.\n{}\n".format(url, resp.status_code, resp.text)
+
+            if resp.status_code == 403:
+                try:
+                    if resp.json()['message'].startswith('API rate limit exceeded') and t < max_retries:
```
Comment (alesapin) [inline utils/make_changelog.py]: `t < max_retries` -- always true?

File: utils/make_changelog.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+            user = users[item['user']]
+            author = u'[{}]({})'.format(user['name'] or user['login'], user['html_url'])
+
+        if cat not in groups:
+            groups[cat] = []
+        groups[cat].append(pattern.format(short_descr, id, link, author))
+
+    texts = []
+    for group, text in groups.items():
+        items = [u'* {}'.format(pr) for pr in text]
```
Comment (alesapin) [inline utils/make_changelog.py]: Seems like `u'str'` doesn't make sense in python3 https://www.python.org/dev/peps/pep-0414/.

File: utils/make_changelog.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+
+    if not is_users_loaded:
+        logging.info('Getting users info using github api.')
+        users = get_users_info(pull_requests, commits_info, token, max_retries, retry_timeout)
+        state['users'] = users
+
+    logging.info('Found {} users.'.format(len(users)))
+    save_state(state_file, state)
+
+    print(process_pull_requests(pull_requests, users, repo))
```
Comment (alesapin) [inline utils/make_changelog.py]: `from __future__ import print_function` add to start of file.

File: utils/make_changelog.py (hunk 7 above)
Comment (alesapin) [reply utils/make_changelog.py]: But you want to support python 2, than OK.

Comment (KochetovNicolai) [reply utils/make_changelog.py]: Yes, now it works for both pythons for me.

**Pull request merged**